monitoring/tests/test_publishers.py

```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env python3
+import unittest
+from unittest.mock import patch
+from sqlalchemy import create_engine
+from sqlalchemy.orm import sessionmaker
+from publishers.mysql_publisher_adapter import MySQLPublisherAdapter
+from publishers.database import Build, Stage, MonitoringBase
+from stats import BuildStats, StageStats
+
+
+class TestConnectionString(unittest.TestCase):
+    def test_get_connection_string(self):
+        connection_str = MySQLPublisherAdapter.get_connection_string('1.2.3.4', 'admin', 'test123', 'TestDB')
+        self.assertEqual(connection_str, 'mysql://admin:test123@1.2.3.4/TestDB')
+
+
+class TestPublishing(unittest.TestCase):
+    def setUp(self):
+        # Provision the SQLite in-memory database with our schema
+        self.engine = create_engine('sqlite://')
+        MonitoringBase.metadata.create_all(self.engine)
+
+        self.session_factory = sessionmaker()
+        self.session_factory.configure(bind=self.engine)
+        self.session = self.session_factory()
+
+    def tearDown(self):
+        self.session = None
+        self.session_factory = None
+        self.engine.dispose()
+        self.engine = None
+
+    def test_publish_build_stats_no_stages(self):
+        build_stats = BuildStats(
+            job_name = 'Test',
+            build_id = 1234,
+            build_url = 'http://localhost:8080/job/MyJob/1',
+            finished_at_secs = 5678,
+            status = 'SUCCESS',
+            duration_millis = 4321,
+            stages = [],
+        )
+
+        with patch.object(MySQLPublisherAdapter, 'get_database_session', return_value=self.session):
```

Comment:
We had some discussion here. get_database_session is an implementation detail of MySQLPublisherAdapter. It should be private.

But then, we mock a private method here (and in other tests). Which kind of smells and suggests that there's another class lurking somewhere (or something else).

Because get_database_session can be parametrized by connection string, maybe we should inject connection_string to MySQLPublisherAdapter and rename it to DatabasePublisherAdapter? A more involved solution would be to implement a MySQLSessionHandler class with get_connection_string and get_database_session method but it seems too complicated for now, so let's not do it.

In any case, _get_database_session should be private, and mocking a private method smells.

Reply:
Actually, why not have an interface called `DatabaseSession`. It would have a method `get_database_session`. 

`DatabaseSession` would be injected to `DatabasePublisherAdapter`.

We would have two implementations: `MySQLSession` and `StubSQLiteSession`. We could use the former in production code and the latter in tests.

Then in the test set up we would have something like:
```
create_engine("sqlite://")
self.session = StubSQLiteSession("sqlite://")
...
```